packages/@cdktf/hcl2cdk/lib/generation.ts

```diff
@@ -904,3 +947,84 @@ export function wrapCodeInConstructor(
     { syntacticPlaceholders: true, plugins: ["typescript"] }
   )({ code, base: baseContainerClass, className: t.identifier(className) });
 }
+
+export const providerConstructImports = (importable: ImportableConstruct[]) => {
+  let provider = importable[0].provider;
+  let namespace = importable[0].namespace;
+  const names = importable.map((i) => i.constructName);
+
+  if (provider === "cdktf" || provider === "constructs") {
+    return template(
+      `import { ${names.join(", ")} } from "${provider}"`
+    )() as t.Statement;
+  }
+
+  if (namespace) {
+    namespace = snakeCase(namespace).replace(/_/g, "-");
+  }
+
+  // Special cases to undo provider names that we override
+  if (provider === "NullProvider") {
+    provider = "null";
+  }
+
+  return template(
+    `import { ${names.join(
+      ", "
+    )} } from "./.gen/providers/${provider}/${namespace}"`
+  )() as t.Statement;
+};
+
+export function buildImports(importables: ImportableConstruct[]) {
+  const groupedImportables = importables.reduce((acc, importable) => {
+    const ns = importable.namespace || "";
+    // Doing some hacky ordering of the imports to make them look a bit nicer
+    const prefix =
```

Comment:
Why not keep sorting separately? and meta information what type it is could be stored/augmented into the list. Might make this less entangled

Reply:
Can you explain a bit more? The original reason for grouping is to get things like:

```ts
import { TerraformVariable, TerraformStack } from 'cdktf';
```

to work, since each importable is added to the array separately. The thing I don't know is whether this is _only_ required for `cdktf` or do we have common namespaces for multiple resources too. If we know for sure we don't need to group anything other than `cdktf`, then we can split those apart.

Reply:
I’m not against combining them, I’m just saying the way you are achieving this goal is a bit convoluted and could be cleaned up by some map, reduce, sort calls. No hard feelings though

Reply:
gotcha. Yeah, I'm sure there's definitely a cleaner way for doing this. That being said, I'd like to get this one done and try to get your imports path fixes rebased on top and see if we're all good there. I'll keep it in mind to think about a revision later